api-samples/tabs/inspector/tabs_api.js

```diff
@@ -0,0 +1,403 @@
+
+  for (let window of windowList) {
+    const windowItem = document.importNode(windowTemplate, true);
+    renderWindow(window, windowItem.children[0]);
+
+    output.appendChild(windowItem);
+  }
+}
+
+function renderWindow(window, windowItem) {
+  windowItem.id = `window_${window.id}`;
+  windowItem.querySelector('.window_left').id = `left_${window.id}`;
+  windowItem.querySelector('.window_top').id = `top_${window.id}`;
+  windowItem.querySelector('.window_width').id = `width_${window.id}`;
+  windowItem.querySelector('.window_height').id = `height_${window.id}`;
+  windowItem.querySelector('.window_focused').id = `focused_${window.id}`;
+  windowItem.querySelector('.window_current').id = `current_${window.id}`;
+  windowItem.querySelector('.window_id').innerText = window.id;
+  windowItem.querySelector('.window_left').value = window.left;
+  windowItem.querySelector('.window_top').value = window.top;
+  windowItem.querySelector('.window_width').value = window.width;
+  windowItem.querySelector('.window_height').value = window.height;
+  windowItem.querySelector('.window_focused').checked = window.focused;
+  windowItem.querySelector('.window_current').checked = window.current;
+
+  windowItem.querySelector('.window_refresh').addEventListener('click', () => {
+    refreshWindow(window.id);
+  });
+
+  windowItem
+    .querySelector('.update_window_button')
+    .addEventListener('click', () => {
+      updateWindow(window.id);
+    });
+
+  windowItem
+    .querySelector('.remove_window_button')
+    .addEventListener('click', () => {
+      removeWindow(window.id);
+    });
+
+  windowItem
+    .querySelector('.refresh_selected_tab_button')
+    .addEventListener('click', () => {
+      refreshSelectedTab(window.id);
+    });
+
+  windowItem.querySelector('#tabList').innerHTML = '';
+  for (let tab of window.tabs) {
+    const tabItem = document.importNode(tabTemplate, true);
+    renderTab(tab, tabItem.children[0]);
+    windowItem.querySelector('#tabList').appendChild(tabItem);
+  }
+}
+
+function renderTab(tab, tabItem) {
+  tabItem.id = `tab_${tab.id}`;
+  tabItem.querySelector('.tab_index').id = `index_${tab.id}`;
+  tabItem.querySelector('.tab_window_id').id = `windowId_${tab.id}`;
+  tabItem.querySelector('.tab_title').id = `title_${tab.id}`;
+  tabItem.querySelector('.tab_url').id = `url_${tab.id}`;
+  tabItem.querySelector('.tab_selected').id = `selected_${tab.id}`;
+
+  tabItem.querySelector('.tab_id').innerText = `TabId: ${tab.id}`;
+  tabItem.querySelector('.tab_index').value = tab.index;
+  tabItem.querySelector('.tab_window_id').value = tab.windowId;
+  tabItem.querySelector('.tab_title').value = tab.title;
+  tabItem.querySelector('.tab_url').value = tab.url;
+  tabItem.querySelector('.tab_selected').checked = tab.selected;
+
+  tabItem.querySelector('.move_tab_button').addEventListener('click', () => {
+    moveTab(tab.id);
+  });
+  tabItem.querySelector('.refresh_tab_button').addEventListener('click', () => {
```

Comment:
It looks like each time we run this we add another listener, which means that if you click "Refresh" tab twice it will actually run three times?